### PR TITLE
librustc_driver => 2018

### PR DESCRIPTION
**Pull request by taiki-e**
Transitions `librustc_driver` to Rust 2018; cc #58099

r? @Centril

### CODE REVIEW

**Review by Centril — APPROVED** [review]
LGTM modulo ICE

**Comment by Centril** [comment]
@taiki-e Try rebasing now :)

**Comment by Centril** [comment]
So... I was hoping that https://github.com/rust-lang/rust/pull/58110/commits/f9e9c9196a196e95d8f39aa57ed30857525be5e6 would have solved the ICE here but apparently not... Could you try this locally with `RUST_BACKTRACE=1 ./x.py test` ?

**Comment by taiki-e** [comment]
@Centril Here is a full dump with backtrace: https://gist.github.com/taiki-e/f034a84e6433a3d3c99aa99354fb4195

**Comment by Centril** [comment]
Seems like the problem occurs here:
```
 15: rustc_resolve::build_reduced_graph::<impl rustc_resolve::Resolver<'a>>::populate_module_if_necessary
```

r? @petrochenkov

**Comment by petrochenkov** [comment]
Minimized reproduction would be appreciated.

File: src/librustc_resolve/build_reduced_graph.rs [hunk 1]
```diff
@@ -679,9 +682,11 @@ impl<'a> Resolver<'a> {
...
                                 (child.def, ty::Visibility::Public, DUMMY_SP, expansion));
 
-                    if self.cstore.associated_item_cloned_untracked(child.def.def_id())
-                           .method_has_self_argument {
-                        self.has_self.insert(child.def.def_id());
+                    if let Some(def_id) = child.def.opt_def_id() {
+                        if self.cstore.associated_item_cloned_untracked(def_id)
+                            .method_has_self_argument {
+                            self.has_self.insert(def_id);
+                        }
```
Comment (petrochenkov) [inline src/librustc_resolve/build_reduced_graph.rs]: I'm pretty sure this shouldn't be necessary for fixing the ICE.

File: src/librustc_resolve/build_reduced_graph.rs [hunk 2]
```diff
@@ -639,7 +639,10 @@ impl<'a> Resolver<'a> {
         // but metadata cannot encode gensyms currently, so we create it here.
         // This is only a guess, two equivalent idents may incorrectly get different gensyms here.
         let ident = ident.gensym_if_underscore();
-        let def_id = def.def_id();
+        let def_id = match def.opt_def_id() {
+            Some(def_id) => def_id,
+            None => return,
```
Comment (petrochenkov) [inline src/librustc_resolve/build_reduced_graph.rs]: Ignoring things without a def-id is not a correct behavior.

`def_id`s should rather be obtained though destructuring inside `match def { ... }`, and new `Def` variants need to be added the match as needed, and they should go through `self.define(...)` too.

File: src/librustc_resolve/build_reduced_graph.rs [hunk 3]
```diff
@@ -701,6 +701,9 @@ impl<'a> Resolver<'a> {
             Def::Macro(..) => {
                 self.define(parent, ident, MacroNS, (def, vis, DUMMY_SP, expansion));
             }
+            Def::NonMacroAttr(..) => {
+                self.define(parent, ident, MacroNS, (def, vis, DUMMY_SP, expansion));
+            }
```
Comment (petrochenkov) [inline src/librustc_resolve/build_reduced_graph.rs]: Nit: This can be merged with the previous match arm.

**Comment by taiki-e** [comment]
@Centril @petrochenkov
`x86_64-gnu-llvm-6.0` passed, but `mingw-check` failed...

Here is a full dump with backtrace of `mingw-check`: https://gist.github.com/taiki-e/52632a55a5f5152da66ce981ffa34ea3

It seems to occur in the same place.

```
  15: rustc_resolve::build_reduced_graph::<impl rustc_resolve::Resolver<'a>>::populate_module_if_necessary
```

**Comment by Dylan-DPC** [comment]
ping from triage @taiki-e any updates?

**Comment by Dylan-DPC** [comment]
ping from triage @taiki-e you have conflicts to resolve

**Comment by taiki-e** [comment]
@Dylan-DPC I'd like to wait for #59047 to land. It will fix the ICE occurred in this PR.

**Comment by Dylan-DPC** [comment]
@talki-e thanks for the reply. Marking this as blocked.

**Comment by petrochenkov** [comment]
https://github.com/rust-lang/rust/pull/59047 has landed, unblocking.

**Comment by Centril** [comment]
@petrochenkov I don't think we changed the bootstrap compiler yet tho.

**Comment by petrochenkov** [comment]
Ok, blocking again.

**Comment by Centril** [comment]
Bootstrap was changed in https://github.com/rust-lang/rust/pull/59434. This should now be unblocked. @taiki-e mind rebasing + other needed changes?

**Comment by petrochenkov** [comment]
@taiki-e 
Could you also squash the commits together with rebase.

**Comment by taiki-e** [comment]
Rebased.

**Pull request merged**